plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgresQueryFailureRecoveryTest.java

```diff
@@ -22,4 +26,22 @@ public TestPostgresQueryFailureRecoveryTest()
     {
         super(RetryPolicy.QUERY);
     }
+
+    @Test
+    @Override
+    protected void testDeleteWithSubquery()
+    {
+        // TODO: support merge with fte https://github.com/trinodb/trino/issues/23345
+        assertThatThrownBy(super::testDeleteWithSubquery).hasMessageContaining("expected: OptionalLong[527]\n but was: OptionalLong[0]");
```

Comment:
Now I got it. It is not that is not supported, it "works" but it returns incorrect results. In other words the MERGE operation for JDBC connectors cannot be retried.

I think what we need to here is to disallow running MERGE in FTE mode for JDBC connectors now. We need to throw good error message, not incorrect results.

Reply:
@losipiuk do you have any idea how connector can recognize if it is executing under FTE mode?

Reply:
> Now I got it. It is not that is not supported, it "works" but it returns incorrect results. In other words the MERGE operation for JDBC connectors cannot be retried.

More Interesting thing is the TASK level retry returns correct results :)

Reply:
There is `verifyRetryMode` in `DefaultJdbcMetadata`, maybe we could put sth in `beginMerge`
```java
if (retryMode != NO_RETRIES) {
     if (!jdbcClient.supportsMergeRetries()) {
           throw new TrinoException(NOT_SUPPORTED, "This connector does not support MERGE with retries");
     }
     if (isNonTransactionalInsert(session)) {
           throw new TrinoException(NOT_SUPPORTED, "Query and task retries are incompatible with non-transactional inserts");
     }
}
``` 
And add the method `supportsMergeRetries` in `JdbcClient`

Reply:
@kokosing Try above, PTAL

Reply:
> There is verifyRetryMode in DefaultJdbcMetadata, maybe we could put sth in beginMerge

Yeah - checking the `retryMode` param in `beginMerge` and acting according to that is the way to go.
I did not read the PR so I do not know if it is supposed to work out of the box with FTE (I doubt it)  or not.